### ISSUE TITLE
C++: Incorrect inheritance graph when using templates

### ISSUE THREAD

**Issue by coolgoose85** [doxygen/doxygen#7897]
**Problem Description**

1. Class inheritance Graph is incorrectly generated  when class inheriting another template class. 
2. Private data member is also not populated for the problem class but for other classes private data members are populated in class graph. (Check screenshot attachments)

In below example _class concreteObserver_ is publicly inheriting _class iObserver_ but Class Hierarchy page do not show it as such.

```
template <typename T>
class iObserver
{
public:
    virtual void update() = 0;
    virtual T getData() = 0;
    virtual void removeObservable(iObservable<T> *cObservableObj) = 0;
};


template <typename T>
class concreteObserver : public iObserver<T>
{
private:
    iObservable<T> *cObservableObj;
    T data;

public:
    concreteObserver() {}
    ~concreteObserver();
    concreteObserver(iObservable<T> *obj);

    void update();
    T getData();
    void removeObservable(iObservable<T> *cObservableObj);
};
```

Also, just moving the class up and down also changes the documentation being generated. I was horrified by this behavior. 

**Expected behavior**
- If one class is inheriting another class it should show it correctly.
- Private data members should be populated in UML class diagram

**Screenshots**
Attached screenshots

**To Reproduce**
Attaching source code and Doxygen config

**Version**
└─ $ ▶ doxygen --version
1.8.18
Using Mac 19F66

**Additional context**
Four classes in the source code
```
template <typename T>
class iObservable;
template <typename T>
class concreteObservable;

template <typename T>
class iObserver;
template <typename T>
class concreteObserver;

```
iObservable <- class concreteObservable

[Problem Relationship]
iObserver <- concreteObserver 
- I would expect class iObserver should show inheritance relationship which is not populated
- I would expect in class diagram for "class concreteObserver" it should populate private data members like it did for "class concreteObservable"
[source_plus_screenshot.zip](https://github.com/doxygen/doxygen/files/4882017/source_plus_screenshot.zip)

**Comment by ohiogeek8215** [comment]
This is all some kind of long setup. This is ridiculous.

**Comment by coolgoose85** [comment]
> This is all some kind of long setup. This is ridiculous.

This is not even a complex project but simple Publisher/Subscriber design pattern. Not sure how is it even close to ridiculous?

**Comment by ohiogeek8215** [comment]
Thanks cool goose!

**Comment by albert-github** [comment]
@coolgoose85 
- Please also provide the information from the doxygen configuration file.

**Comment by coolgoose85** [comment]
Attaching Doxygen Config file.
[CodeRepository_deoxygen.txt](https://github.com/doxygen/doxygen/files/4885982/CodeRepository_deoxygen.txt)

**Comment by coolgoose85** [comment]
> @coolgoose85
> 
> * Please also provide the information from the doxygen configuration file.

Attached Doxygen config file above.

**Comment by albert-github** [comment]
A quick first guess is that it is related to the, forward declaration, lines:
```
template <typename T>
class iObserver;
template <typename T>
class concreteObserver;    
```
in the file observerPattern.hpp (lines 17 - 20)

**Comment by coolgoose85** [comment]
> A quick first guess is that it is related to the, forward declaration, lines:
> 
> ```
> template <typename T>
> class iObserver;
> template <typename T>
> class concreteObserver;    
> ```
> 
> in the file observerPattern.hpp (lines 17 - 20)

I have tried below cases and they all generate different data by Doxygen:

Case 1
Forward declare only Class iObserver and Class concreteObserver
- Class iObserver needs to be forward declared as it is used by Class iObservable
- Class concreteObserver does not needs to be forward declared.


Actually I just tried that i.e. remove Class concreteObserver from forward declaration and **it worked**.

Seems like compiler is forgiving of it but Doxygen is not forgiving of it. Can you please point me to any documentation which goes over Dos/Don't and pitfalls.

In this particular case, I can technically forward declare all the classes which I did at one time and compiler compiles it fine because it is just an indication to compiler that classes are coming but Doxygen in that case did not populate any inheritance graphs.

Apparently, Doxygen gets either confused or does strict enforcement where is needs to see only forward declaration of classes which are actually required which is this case.

**Comment by albert-github** [comment]
There is, to the best of my knowledge, no document with Dos/Don't and pitfalls.

I think in this case you could place the mentioned part in a `#ifndef` construct like:
```
#ifndef RUN_DOXYGEN
template <typename T>
class iObserver;
template <typename T>
class concreteObserver; 
#endif
```
and `PREDEFINED` `RUN_DOXYGEN` in your doxygen configuration file. 

This is not a solution, but just a work around.

**Comment by coolgoose85** [comment]
> There is, to the best of my knowledge, no document with Dos/Don't and pitfalls.
> 
> I think in this case you could place the mentioned part in a `#ifndef` construct like:
> 
> ```
> #ifndef RUN_DOXYGEN
> template <typename T>
> class iObserver;
> template <typename T>
> class concreteObserver; 
> #endif
> ```
> 
> and `PREDEFINED` `RUN_DOXYGEN` in your doxygen configuration file.
> 
> This is not a solution, but just a work around.

Are you saying Doxygen do not like forward declaration and with ifndef you can guide Doxygen to skip skimming through forward declaration? Is that the correct understanding?

**Comment by albert-github** [comment]
It looks indeed like doxygen does not like the forward declarations, at the moment, and indeed with the #ifndef construct and setting you can hide the mentioned part of the code from doxygen.

**Comment by albert-github** [comment]
I think by the fix of #8177 the problem here is fixed (but not released) as well (same problem in my opinion).

- @coolgoose85 Please test you problem against the current master.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
